File: scripts/fetchComments.ts
import { ThreadIdFetchError, ThreadKeyFetchError } from "./error.ts";

interface ThreadRequestBody {
    params: {
        targets: {
            id: string;
            fork: "owner" | "main" | "easy";
        }[];
        language: "ja-jp";
    };
    threadKey: string;
    additionals: Record<string | number | symbol, never>;
}

interface ThreadResponse {
    meta: {
        status: number;
    };
    data: {
        globalComments: {
            id: string;
            count: number;
        }[];
        threads: {
            id: number;
            fork: "owner" | "main" | "easy";
            commentCount: number;
            comments: {
                id: string;
                no: number;
                vposMs: number;
                body: string;
                commands: string[];
                userId: string;
                isPremium: boolean;
                score: number;
                postedAt: string;
                nicoruCount: number;
                nicoruId: string | null;
                source: "trunk" | "leaf" | "nicoru";
                isMyPost: boolean;
            }[];
        }[];
    };
}

// 入力されたsm番号からニコニコ動画のコメントを取得する
export async function fetchComments(videoId: string): Promise<void> {
    const url = `https://www.nicovideo.jp/watch/${videoId}`;
    const endpoint = "https://public.nvcomment.nicovideo.jp/v1/threads";
    const threadIdRegex = /threadIds&quot;:\[\{&quot;id&quot;:(.*?),&quot;/;
    const threadKeyRegex = /{&quot;threadKey&quot;:&quot;(eyJ0eXAiOiJKV1Qi.*?)&quot/;

    // 動画ページを取得
    const videoPage = await (await fetch(url)).text();
    const threadId = videoPage.match(threadIdRegex)?.[1];
    const threadKey = videoPage.match(threadKeyRegex)?.[1];

    // スレッドIDまたはスレッドキーが取得できない場合のエラー
    if (!threadId) {
        throw new ThreadIdFetchError("スレッドIDの取得に失敗しました。");
    }
    if (!threadKey) {
        throw new ThreadKeyFetchError("スレッドキーの取得に失敗しました。");
    }

    // リクエストペイロード
    const payload: ThreadRequestBody = {
        params: {
            targets: [
                { id: threadId, fork: "owner" },
                { id: threadId, fork: "main" },
                { id: threadId, fork: "easy" },
            ],
            language: "ja-jp",
        },
        threadKey: threadKey,
        additionals: {},
    };

    // コメント取得APIを呼び出し
    const response = await fetch(endpoint, {
        method: "POST",
        body: JSON.stringify(payload),
        headers: { "x-frontend-id": "6" },
    });
    const data = (await response.json()) as ThreadResponse;

    // コメントを表示
    data.data.threads.forEach((thread) => {
        console.log(`コメントタイプ: ${thread.fork}`);
        thread.comments.forEach((comment) => {
            const time = `${Math.floor(comment.vposMs / 60000)}:${Math.floor((comment.vposMs % 60000) / 1000)}`;
            console.log(`[${time}] ${comment.body}`);
        });
    });
}

// GitHub Actions から動画IDを渡す
const videoId = Deno.args[0];
if (videoId) {
    fetchComments(videoId).catch((error) => {
        console.error("エラーが発生しました:", error);
    });
} else {
    console.error("動画IDが指定されていません。");
}

File: scripts/fetch_comments.py
import re
import requests
import sys
import os

class ThreadIdFetchError(Exception):
    pass

class ThreadKeyFetchError(Exception):
    pass

def fetch_comments(video_id: str, output_file: str):
    url = f"https://www.nicovideo.jp/watch/{video_id}"
    endpoint = "https://public.nvcomment.nicovideo.jp/v1/threads"
    thread_id_regex = re.compile(r'threadIds&quot;:\[\{&quot;id&quot;:(.*?),&quot;')
    thread_key_regex = re.compile(r'{&quot;threadKey&quot;:&quot;(eyJ0eXAiOiJKV1Qi.*?)&quot')

    # 動画ページのHTMLを取得
    response = requests.get(url)
    if response.status_code != 200:
        raise Exception(f"Failed to fetch video page. Status code: {response.status_code}")
    video_page = response.text

    # threadId と threadKey を抽出
    thread_id_match = thread_id_regex.search(video_page)
    thread_key_match = thread_key_regex.search(video_page)
    if not thread_id_match:
        raise ThreadIdFetchError("Thread ID not found.")
    if not thread_key_match:
        raise ThreadKeyFetchError("Thread Key not found.")

    thread_id = thread_id_match.group(1)
    thread_key = thread_key_match.group(1)

    # コメント取得リクエストのペイロード
    payload = {
        "params": {
            "targets": [
                {"id": thread_id, "fork": "owner"},
                {"id": thread_id, "fork": "main"},
                {"id": thread_id, "fork": "easy"}
            ],
            "language": "ja-jp"
        },
        "threadKey": thread_key,
        "additionals": {}
    }

    # コメント取得リクエストを送信
    headers = {"x-frontend-id": "6"}
    thread_response = requests.post(endpoint, json=payload, headers=headers)
    if thread_response.status_code != 200:
        raise Exception(f"Failed to fetch comments. Status code: {thread_response.status_code}")

    comments = thread_response.json()

    # 出力先ディレクトリを作成（存在しない場合）
    os.makedirs(os.path.dirname(output_file), exist_ok=True)

    # コメントをファイルに保存
    with open(output_file, "w", encoding="utf-8") as f:
        for thread in comments.get("data", {}).get("threads", []):
            for comment in thread.get("comments", []):
                try:
                    time = float(comment.get("vposMs", 0)) / 1000  # 秒単位に変換
                    body = comment.get("body", "").replace("\n", " ")  # 改行を除去
                    if body:  # コメントが空でない場合のみ保存
                        f.write(f"{time} {body}\n")
                except (ValueError, TypeError):
                    # vposMs が無効な値の場合はスキップ
                    continue

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python fetch_comments.py <video_id>")
        sys.exit(1)

    video_id = sys.argv[1]
    output_file = f"downloads/{video_id}_comments.txt"
    try:
        fetch_comments(video_id, output_file)
        print(f"Comments saved to {output_file}")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)
